### PR TITLE
Cache index queries for up to 15mins

### PR DESCRIPTION
**Pull request by tomwilkie**
An in-process cache for index queries to accelerate queries.

Cache entries are valid for 15mins as thats how long the ingester will hold flushed chunks for.

### CODE REVIEW

**Comment by bboreham** [comment]
> Cache entries are valid for 15mins as thats how long the ingester will hold flushed chunks for.

Is that a recent change? How does that relate to https://github.com/weaveworks/cortex/blob/master/pkg/chunk/chunk_store.go#L208 ?

**Comment by tomwilkie** [comment]
We've got min-chunk-age set to 15mins in our cluster.  

The challenging with caching index entries is that that we need to ensure we don't miss a series, even when querying more that 15 mins in the past.  So knowing that any entries in the cache are at max 15mins old, and knowing that the ingester is going to response with anything more recent than 15mins, allows us to make this optimisation.  To be clear: this is a lower bound on the the cache freshness; we can probably cache most stuff for much longer (anything read from yesterdays rows can be cached for ever).

A counter example might be useful.  Imagine:
- t=0: we write a sample for `foo{}` to the ingesters
- t=10: we do a query for `foo{}[1h]` and find no series in the index, but get the response from the ingester
- t=15: `foo{}` is flushed
- t=20: we do a query for `foo{}[1h]` and find (using the cache) no series in the index, but get the response from the ingester
- t=30: `foo{}` is dropped from the ingester
- t=35: we do a query for `foo{}[1h]` and find (using the cache) no series in the index, and get no response from the ingester <- wrong query response.

Hence, we much drop the cache entries after 15mins.  Make sense?

**Comment by tomwilkie** [comment]
> Does it become consistent if we add a line in the documentation stating that -store.index-cache-validity should be no higher than -ingester.max-chunk-idle ?

Yes we should add that line for sure.

> Separately, I would be very keen to cache index lookups from previous days' buckets.

Agreed!  Should be relatively easy to do after this change.  We've had it as a TODO in #209 for 18 months!

File: pkg/chunk/storage/factory.go [hunk 1]
```diff
@@ -28,26 +32,32 @@ func (cfg *Config) RegisterFlags(f *flag.FlagSet) {
 	cfg.AWSStorageConfig.RegisterFlags(f)
 	cfg.GCPStorageConfig.RegisterFlags(f)
 	cfg.CassandraStorageConfig.RegisterFlags(f)
+
+	f.IntVar(&cfg.IndexCacheSize, "store.index-cache-size", 0, "Size of in-memory index cache, 0 to disable.")
+	f.DurationVar(&cfg.IndexCacheValidity, "store.index-cache-validity", 15*time.Minute, "Period for which entries in the index cache are valid. Should be no higher than -ingester.max-chunk-idle.")
```
Comment (csmarchbanks) [inline pkg/chunk/storage/factory.go]: The default here needs to match the default for `-ingester.max-chunk-idle`, which is 5 minutes right now

Comment (gouthamve) [reply pkg/chunk/storage/factory.go]: Yup sorry.

**Comment by tomwilkie** [comment]
This is running and has made a big difference for us; LGTM from me.

File: pkg/chunk/storage/factory.go [hunk 2]
```diff
@@ -28,26 +32,32 @@ func (cfg *Config) RegisterFlags(f *flag.FlagSet) {
...
+		client, err = aws.NewStorageClient(cfg.AWSStorageConfig, schemaCfg)
 	case "gcp":
-		return gcp.NewStorageClient(context.Background(), cfg.GCPStorageConfig, schemaCfg)
+		client, err = gcp.NewStorageClient(context.Background(), cfg.GCPStorageConfig, schemaCfg)
 	case "cassandra":
-		return cassandra.NewStorageClient(cfg.CassandraStorageConfig, schemaCfg)
+		client, err = cassandra.NewStorageClient(cfg.CassandraStorageConfig, schemaCfg)
 	default:
-		return nil, fmt.Errorf("Unrecognized storage client %v, choose one of: aws, gcp, cassandra, inmemory", cfg.StorageClient)
+		client, err = nil, fmt.Errorf("Unrecognized storage client %v, choose one of: aws, gcp, cassandra, inmemory", cfg.StorageClient)
```
[the remark this answers is not in the document]
Comment (gouthamve) [reply pkg/chunk/storage/factory.go]: Done.

File: pkg/chunk/fifo_cache.go [hunk 3]
```diff
@@ -1,20 +1,79 @@
 package chunk
```
Comment (csmarchbanks) [inline pkg/chunk/fifo_cache.go]: Mostly a note/question, it doesn't look like FifoCache is related to the chunk. Could it be pulled out to a different package in the future?

File: pkg/chunk/storage/caching_storage_client.go [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+
+func newCachingStorageClient(client chunk.StorageClient, size int, validity time.Duration) chunk.StorageClient {
+	if size == 0 {
+		return client
+	}
+
+	return &cachingStorageClient{
+		StorageClient: client,
+		cache:         chunk.NewFifoCache("index", size, validity),
+		validity:      validity,
```
Comment (csmarchbanks) [inline pkg/chunk/storage/caching_storage_client.go]: Is validity needed on the `cachingStorageClient`? Or only to create the cache?

**Comment by csmarchbanks** [comment]
A couple final small questions/comments. Then this LGTM.

File: pkg/chunk/storage/caching_storage_client.go (hunk 4 above)
Comment (gouthamve) [reply pkg/chunk/storage/caching_storage_client.go]: You're right. It was done outside the cache before, now I've pushed it into the cache. Forgot to remove this.

**Comment by gouthamve** [comment]
@csmarchbanks fixed all comments, thanks!

**Pull request merged**